### ISSUE TITLE
Torch.Autograd.Backward?

### ISSUE THREAD

**Issue by GeorgeS2019** [dotnet/TorchSharp#691]
It seems [Torch.Tensor.Backward](https://pytorch.org/docs/stable/generated/torch.Tensor.backward.html) is implemented in **Tensor.cs** but not the [`Torch.Autograd.Backward`](https://pytorch.org/docs/stable/generated/torch.autograd.backward.html) within **Autograd.cs**?

Part of [Autograd discussions](https://github.com/dotnet/TorchSharp/issues/606)

**Comment by NiklasGustafsson** [comment]
Straight-forward to implement. Is there a good example that i can use for unit testing?

**Comment by GeorgeS2019** [comment]
I found this, I hope it has the relevant tests for [Torch.Autograd.Backward](https://pytorch.org/docs/stable/generated/torch.autograd.backward.html)

https://github.com/python-engineer/pytorchTutorial/blob/master/03_autograd.py

```python
import torch
# The autograd package provides automatic differentiation 
# for all operations on Tensors

# requires_grad = True -> tracks all operations on the tensor. 
x = torch.randn(3, requires_grad=True)
y = x + 2

# y was created as a result of an operation, so it has a grad_fn attribute.
# grad_fn: references a Function that has created the Tensor
print(x) # created by the user -> grad_fn is None
print(y)
print(y.grad_fn)

# Do more operations on y
z = y * y * 3
print(z)
z = z.mean()
print(z)

# Let's compute the gradients with backpropagation
# When we finish our computation we can call .backward() and have all the gradients computed automatically.
# The gradient for this tensor will be accumulated into .grad attribute.
# It is the partial derivate of the function w.r.t. the tensor

z.backward()
print(x.grad) # dz/dx

# Generally speaking, torch.autograd is an engine for computing vector-Jacobian product
# It computes partial derivates while applying the chain rule
```

**Comment by GeorgeS2019** [comment]
### [Compute gradient with autograd](https://github.com/python-engineer/pytorchTutorial/blob/master/05_2_gradientdescent_auto.py)


```python
import torch

# Here we replace the manually computed gradient with autograd

# Linear regression
# f = w * x 

# here : f = 2 * x
X = torch.tensor([1, 2, 3, 4], dtype=torch.float32)
Y = torch.tensor([2, 4, 6, 8], dtype=torch.float32)

w = torch.tensor(0.0, dtype=torch.float32, requires_grad=True)

# model output
def forward(x):
    return w * x

# loss = MSE
def loss(y, y_pred):
    return ((y_pred - y)**2).mean()

print(f'Prediction before training: f(5) = {forward(5).item():.3f}')

# Training
learning_rate = 0.01
n_iters = 100

for epoch in range(n_iters):
    # predict = forward pass
    y_pred = forward(X)

    # loss
    l = loss(Y, y_pred)

    # calculate gradients = backward pass
    l.backward()

    # update weights
    #w.data = w.data - learning_rate * w.grad
    with torch.no_grad():
        w -= learning_rate * w.grad
    
    # zero the gradients after updating
    w.grad.zero_()

    if epoch % 10 == 0:
        print(f'epoch {epoch+1}: w = {w.item():.3f}, loss = {l.item():.8f}')

print(f'Prediction after training: f(5) = {forward(5).item():.3f}')
```

**Comment by NiklasGustafsson** [comment]
I don't see an example of autograd.backward in either one of those two examples. I'll see what I can find online.
